File: python/2150/main.py
from collections import deque
MAX = 10001

v, e = map(int, input().split())

graph = [[] for _ in range(MAX)]
finished = [False] * MAX
parent = [0] * MAX
stack = deque()
SCC = []
id = 1


def dfs(start):
    global id
    parent[start] = id
    id += 1
    stack.append(start)

    result = parent[start]
    for will in graph[start]:
        if parent[will] == 0:
            result = min(result, dfs(will))
        elif not finished[will]:
            result = min(result, parent[will])

    if parent[start] == result:
        scc = []
        while True:
            top = stack.pop()
            scc.append(top)
            finished[top] = True
            if top == start:
                break
        SCC.append(sorted(scc))
    return result


for _ in range(e):
    a, b = map(int, input().split())
    graph[a].append(b)

for i in range(v):
    if parent[i] == 0:
        dfs(i)
print(len(SCC) - 1)
for scc in sorted(SCC[1:]):
    print(' '.join(str(comp) for comp in scc),-1)

File: Java/2150/src/Main.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();
    static List<List<Integer>> graph;
    static List<List<Integer>> rGraph;
    static List<List<Integer>> SCC;
    static Stack<Integer> stack;
    static boolean[] visited;
    static int V;
    static int E;

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());

        graph = new ArrayList<>();
        rGraph = new ArrayList<>();
        SCC = new ArrayList<>();

        for (int i = 0; i <= V; i++) {
            graph.add(new ArrayList<>());
            rGraph.add(new ArrayList<>());
            SCC.add(new ArrayList<>());
        }

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());

            int now = Integer.parseInt(st.nextToken());
            int will = Integer.parseInt(st.nextToken());
            graph.get(now).add(will);
            rGraph.get(will).add(now);
        }
        visited = new boolean[V + 1];
        stack = new Stack<>();

        for (int i = 1; i < V; i++) {
            if(!visited[i]){
                dfs(i);
            }
        }

        Arrays.fill(visited, false);

        int groupNum = 0;

        while(!stack.isEmpty()){
            int start = stack.pop();

            if(visited[start]){
                continue;
            }

            redfs(start, groupNum++);
        }

        TreeMap<Integer, Integer> tm = new TreeMap<>();

        for (int i = 0; i < groupNum; i++) {
            Collections.sort(SCC.get(i));
            tm.put(SCC.get(i).get(0), i);
        }

        sb.append(groupNum).append("\n");
        tm.keySet().forEach((key) -> {
            int val = tm.get(key);
            for (int will : SCC.get(val)) {
                sb.append(will + " ");
            }
            sb.append("-1\n");
        });

        System.out.println(sb.toString());
    }

    static void dfs(int start){
        visited[start] = true;

        for (int will : graph.get(start)) {
            if(!visited[will]){
                dfs(will);
            }
        }
        stack.push(start);
    }

    static void redfs(int start, int groupNum){
        visited[start] = true;
        SCC.get(groupNum).add(start);

        for (int will : rGraph.get(start)) {
            if(!visited[will]){
                redfs(will, groupNum);
            }
        }
    }
}